js/adapt-contrib-trickle.js

```diff
@@ -1,134 +1,472 @@
+      if (inheritedConfigId) {
+        const inheritedConfig = Adapt.findById(inheritedConfigId).get('_trickle');
+        const localOverrides = model.get('_trickle');
+        const derivedConfig = $.extend(true, {}, inheritedConfig, localOverrides);
+        return derivedConfig;
+      }
       return model.get('_trickle');
-    },
+    }
+
+    /**
+     * Set the specified trickle configuration for the give model
+     * @param {Backbone.Model} model
+     * @param {object} config
+     */
+    setModelConfig(model, config) {
+      model.unset('_trickleConfigId');
+      model.set('_trickle', config);
+    }
+
+    onDataReady() {
+      this.setupEventListeners();
+      this.addComponents();
+      this.applyLocks();
+    }
+
+    setupEventListeners() {
+      this.listenTo(Adapt.data, {
+        'change:_isInteractionComplete': this.applyLocks
+      });
+      this.listenTo(Adapt, {
+        'contentObjectView:preRender': this.applyLocks,
+        'view:addChild': this.onAddChild
+      });
+    }
+
+    /**
+     * Adds trickle button components in their relevant places throughout the course
+     */
+    addComponents() {
+      // Fetch the component model from the store incase it needs overriding by another extension
+      const TrickleModel = Adapt.getModelClass('trickle-button');
+      let uid = 0;
+      Adapt.data.forEach(siteModel => {
+        if (siteModel instanceof CourseModel) return;
+        let trickleConfig = siteModel.get('_trickle');
+        if (!trickleConfig || !trickleConfig._isEnabled) return;
+        // Add a trickle button component model to each trickle site where applicable
+        const siteId = siteModel.get('_id');
+        const buttonModelSites = trickleConfig._onChildren ? siteModel.getChildren().models : [siteModel];
+        buttonModelSites.forEach(buttonModelSite => {
+          buttonModelSite.set({
+            _isTrickled: true,
+            _trickleConfigId: siteId
+          });
+          const trickleModel = new TrickleModel({
+            _id: `trickle-${uid++}`,
+            _type: 'component',
+            _component: 'trickle-button',
+            _parentId: buttonModelSite.get('_id'),
+            _isAvailable: true,
+            _layout: 'full',
+            _trickleConfigId: siteId,
+            _renderPosition: 'outer-append'
+          });
+          trickleModel.setupModel();
+          // Typically this line would append a trickle button to an article or block
+          buttonModelSite.getChildren().add(trickleModel);
+        });
+      });
+    }
 
-    getCompletionAttribute: function() {
-      var trickle = this.getModelConfig(Adapt.config);
+    /**
+     * Calculates and applies the new locking state for each sub-page model according
+     * to each trickle configuration
+     */
+    applyLocks() {
+      if (Adapt.trickle.isKilled) return;
+      const completionAttribute = this.getCompletionAttribute();
+      const locks = {};
```

Comment:
these empty objects "locks" and "modelsById" _could_ be set inside the initialize function.

Reply:
These variables do not reflect the global state of trickle.
They are not used outside of the functions in which they are created.
It would provide no extra clarity or performance having them stored outside of the function.

They could be set inside the initialize function but I don't understand why they should/would be.